doc/OnlineDocs/contribution_guide.rst

```diff
@@ -59,6 +59,128 @@ reviewed or merged by the core development team. In addition, any
+
+We also recommend development be done on branches:
+
+::
+
+   cd pyomo/
+   git branch <branch_name>
+   git checkout <branch_name>
+
+Developement can then be performed. In order to push your branch to your fork, use
+
+::
+
+   git add <filename>
+   git status  # to check that you have added the correct files
+   git commit -m 'informative commit message to describe changes'
+   git push origin <branch_name>
+
+When your changes are ready for a pull request,
+
+    * visit https://github.com/<username>/pyomo.
+    * Just above the list of files and directories in the repository,
+      you should see a button that says "Branch: master". Click on
+      this button, and choose the correct branch.
+    * Click the "New pull request" button just to the right of the
+      "Branch: <branch_name>" button.
+    * Fill out the pull request template and click the green "Create
+      pull request" button.
+
+Merging the Pyomo master branch into your fork/branch
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We recommend updating your fork of Pyomo with changes to Pyomo master
+on a regular basis. This can be done either through GitHub or through
+the command line.
+
+Using GitHub to update your fork
+********************************
+
+    * visit https://github.com/Pyomo/pyomo.
+    * Click on the "New pull request" button just above the list of
+      files and directories.
+    * You will see the title "Compare changes" with some small text
+      below it which says "Compare changes across branches, commits,
+      tags, and more below. If you need to, you can also compare
+      across forks." Click the last part of this: "compare across
+      forks".
+    * You should now see four buttons just below this: "base
+      repository: Pyomo/pyomo", "base: master", "head repository:
+      Pyomo/pyomo", and "compare: master". Click the leftmost button
+      and choose "<username>/Pyomo".
+    * Then click the button which is second to the left, and choose
+      the branch which you want to merge Pyomo master into. The four
+      buttons should now read: "base repository: <username>/pyomo",
+      "base: <branch_name>", "head repository: Pyomo/pyomo", and
+      "compare: master".
+    * You should also now see a pull request template. If you fill out
+      the pull request template and click "Create pull request", this
+      will create a pull request which will update your fork and
+      branch with any changes that have been made to the master branch
+      of Pyomo.
+    * You can then merge the pull request by clicking the green "Merge
+      pull request" button.
+
+Using git commands to update your fork
+**************************************
+
+First, suppose you have set up your remotes with the following:
+
+::
+   
+   git clone https://github.com/<username>/pyomo.git
+   git remote rename origin my_fork
+   git remote add main_fork https://github.com/pyomo/pyomo.git
+
```

Comment:
Is the main Pyomo repository really a "fork"? Is there a better name for this? I could see people thinking that the "main_fork" is their fork.

Reply:
Also, while I agree with this renaming, it does complicate the use of other instructions from the web which would likely use the "origin" and "upstream" names. (Maybe this is better - I don't like the origin and upstream names, because "origin" to me feels like it refers to the main Pyomo repository - when it is really the "upstream")... Other thoughts on naming? I use my_fork and main_pyomo. But I could be convinced otherwise.

Reply:
FWIW, I have settled on `origin` for `Pyomo/pyomo`, `me` for `jsiirola/pyomo`, and first names for everyone else's forks.

Reply:
But then, isn't your "origin" exactly the reverse of the typical instructions online? Where "origin" is your fork (what you call me), and "upstream" is the main repos (what you call "origin")?

Reply:
I have no real preference on the naming.

Reply:
I personally use John's approach but I agree with Carl that using the `origin` name in a slightly different way from other online documentation would likely cause confusion. I like the idea of calling it `main_pyomo`.

Reply:
We're going headlong into a messy developer preference case, but the preponderance of online documentation has `origin` be the user's fork and `upstream` refer to the... well, upstream.

https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow

I don't have an issue with individual developers choosing something different from the mainstream, but since we don't want to have to field all the questions, we probably don't want to publicize a practice deviating from the "I Googled it" answer.

Reply:
I agree with @qtothec and @carldlaird that for the purposes of this documentation, our recommendations should at least be not inconsistent with "I Googled it" (either rename both remotes like @michaelbynum is doing now, or keep them with the standard `origin` and `upstream`).